packages/actions/ezrassor_teleop_actions/scripts/teleop_action_client.py

```
@@ -0,0 +1,88 @@
+#!/usr/bin/env python
+
+import sys
+import rospy
+import actionlib
+
+from ezrassor_teleop_msgs.msg import TeleopAction
+from ezrassor_teleop_msgs.msg import TeleopGoal
+
+
+class TeleopActionClient:
+    def __init__(self):
+        # ROS Actionlib Library
+        self._ac = actionlib.SimpleActionClient(
+            "/teleop_action_server", TeleopAction
+        )
+
+        self._ac.wait_for_server()
+        rospy.loginfo("Connected to Teleop Action Server.")
+
+    def send_movement_goal(self, actions):
+        """Accepts a list of actions to send to the action server.
+        Feedback returned indicates the robot's current x, y, and heading."""
+        for action in actions:
+            instruction = action.split(" ")
+            direction = instruction[0]
+            duration = float(instruction[1])
+
+            goal = TeleopGoal(direction, duration)
+            self._ac.send_goal(
+                goal,
+                done_cb=self.done_callback,
+                feedback_cb=self.feedback_callback,
+            )
+
+            rospy.loginfo("{} has been sent.".format(direction))
+
+            success = self._ac.wait_for_result(rospy.Duration(duration + 2.0))
+            if not success:
+                rospy.logwarn("HOUSTON, WE HAVE A PROBLEM")
+                break
+
+            rospy.loginfo(self._ac.get_result())
+
+    def done_callback(self, status, result):
+        """When the goal is marked as finished, log the info to console."""
+        rospy.loginfo("Status :")
+        rospy.loginfo(str(status))
+
+        rospy.loginfo("Result: ")
+        rospy.loginfo(str(result))
+
+    def feedback_callback(self, feedback):
+        """Feedback comes formatted from the action server."""
+        rospy.loginfo(feedback)
+
+
+def read_instructions(file):
```

Looks like there's no validation on the file, what happens if the 4th instruction is spelled wrong or the number they include is a float? It won't fail fast, it will process the first three instructions and then spin out.

I'd recommend trying to return a list of TeleopGoals from `read_instructions`, but make sure each TeleopGoal is valid. Maybe rosactions supports using an enum type thing for `direction`? Ideally if the file is invalid we would fail in this function, not later in the code.

---

You are 100% right and I just added a validate() function for this purpose.